pkg/infra/db/dbconn/dbconn.go

```diff
@@ -0,0 +1,380 @@
+package dbconn
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/go-sql-driver/mysql"
+	"github.com/grafana/grafana/pkg/infra/fs"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/infra/tracing"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/services/sqlstore/migrator"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/util"
+	"xorm.io/xorm"
+)
+
+var tlslog = log.New("tls_mysql")
+
+type Conn struct {
+	log log.Logger
+
+	dbCfg  databaseConfig
+	engine *xorm.Engine
+	tracer tracing.Tracer
+}
+
+func New(cfg *setting.Cfg, engine *xorm.Engine, tracer tracing.Tracer) (*Conn, error) {
+	// This change will make xorm use an empty default schema for postgres and
+	// by that mimic the functionality of how it was functioning before
+	// xorm's changes above.
+	xorm.DefaultPostgresSchema = ""
+	conn := &Conn{
+		log: log.New("sqlstore"),
+	}
+	err := conn.initEngine(cfg, engine)
+	if err != nil {
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+func (ss *Conn) Engine() *xorm.Engine { return ss.engine }
+
+func (ss *Conn) buildExtraConnectionString(sep rune) string {
+	if ss.dbCfg.UrlQueryParams == nil {
+		return ""
+	}
+
+	var sb strings.Builder
+	for key, values := range ss.dbCfg.UrlQueryParams {
+		for _, value := range values {
+			sb.WriteRune(sep)
+			sb.WriteString(key)
+			sb.WriteRune('=')
+			sb.WriteString(value)
+		}
+	}
+	return sb.String()
+}
+
+func (ss *Conn) buildConnectionString(cfg *setting.Cfg) (string, error) {
+	dbCfg, err := readConfig(cfg)
+	if err != nil {
+		return "", err
+	}
+	ss.dbCfg = dbCfg
```

Comment:
should we separate the cfg to dbcfg mapping out so as much of this code as possible uses a databaseConfig instead of setting.Cfg?